src/main/java/edu/wpi/grip/core/Connection.java

```diff
@@ -29,6 +30,10 @@ public Connection(EventBus eventBus, Socket<? extends T> outputSocket, Socket<T>
         checkNotNull(inputSocket);
         checkNotNull(outputSocket);
         checkNotNull(eventBus);
+        checkArgument(!Socket.Direction.INPUT.equals(outputSocket.getDirection()),
+                "outputSocket cannot be an input socket");
+        checkArgument(!Socket.Direction.OUTPUT.equals(inputSocket.getDirection()),
+                "inputSocket cannot be an output socket");
```

Comment:
Alternatively you could explicitly define a class called `InputSocket` and `OutputSocket`. Then you could make a connection only support one or the other. This would provide compile time safety.